processing/src/test/java/org/apache/druid/query/rowsandcols/semantic/FramedOnHeapAggregatableTest.java

```diff
@@ -456,4 +459,140 @@ public void testReverseCumulativeAggregation()
+    int[] c2Vals = new int[] {0, 1, 1, 1, 3, 4, 5};
+    int[] resVals = new int[] {0, 1, 3, 3, 5, 7, 9};
+
+    simpleWindowingTest(frame, c1Vals, c2Vals, resVals);
+  }
+
+  @Test
+  public void testRangeA1()
+  {
+    WindowFrame frame = new WindowFrame(
+        PeerType.RANGE,
+        false,
+        0,
+        false,
+        1,
+        Collections.singletonList(ColumnWithDirection.ascending("c1"))
+    );
+
+    int[] c1Vals = new int[] {0, 1, 2, 2, 3, 4, 5};
+    int[] c2Vals = new int[] {0, 1, 1, 1, 3, 4, 5};
+    int[] resVals = new int[] {1, 3, 5, 5, 7, 9, 5};
+
+    simpleWindowingTest(frame, c1Vals, c2Vals, resVals);
+  }
+
+  @Test
+  public void testRangeB1A1()
+  {
+    WindowFrame frame = new WindowFrame(
+        PeerType.RANGE,
+        false,
+        1,
+        false,
+        1,
+        Collections.singletonList(ColumnWithDirection.ascending("c1"))
+    );
+
+    int[] c1Vals = new int[] {0, 1, 2, 3, 4, 5};
+    int[] c2Vals = new int[] {0, 1, 2, 3, 4, 5};
+    int[] resVals = new int[] {1, 3, 6, 9, 12, 9};
+
+    simpleWindowingTest(frame, c1Vals, c2Vals, resVals);
+  }
+
+
+  @Test
+  public void testRangeB1A1_2()
+  {
+    WindowFrame frame = new WindowFrame(
+        PeerType.RANGE,
+        false,
+        1,
+        false,
+        1,
+        Collections.singletonList(ColumnWithDirection.ascending("c1"))
+    );
+
+    int[] c1Vals = new int[] {0, 0, 1, 2, 3, 3, 4, 4, 5};
+    int[] c2Vals = new int[] {0, 0, 1, 2, 2, 1, 2, 2, 5};
+    int[] resVals = new int[] {1, 1, 3, 6, 9, 9, 12, 12, 9};
+
+    simpleWindowingTest(frame, c1Vals, c2Vals, resVals);
+  }
+
+  @Test
+  public void testRangeB1A2()
+  {
+    WindowFrame frame = new WindowFrame(
+        PeerType.RANGE,
+        false,
+        1,
+        false,
+        2,
+        Collections.singletonList(ColumnWithDirection.ascending("c1"))
+    );
+
+    int[] c1Vals = new int[] {0, 0, 0, 1, 1, 1, 2, 2, 2, 2, 3, 3, 3};
+    int[] c2Vals = new int[] {1, 1, 2, 1, 1, 2, 1, 1, 1, 2, 1, 1, 1};
+    int[] resVals = new int[] {13, 13, 13, 16, 16, 16, 12, 12, 12, 12, 8, 8, 8};
+
+    simpleWindowingTest(frame, c1Vals, c2Vals, resVals);
+  }
+
+
+  private void simpleWindowingTest(WindowFrame frame, int[] c1Vals, int[] c2Vals, int[] resVals)
+  {
+    Map<String, Column> map = new LinkedHashMap<>();
+    map.put("c1", new IntArrayColumn(c1Vals));
+    map.put("c2", new IntArrayColumn(c2Vals));
+
+    RowsAndColumns rac = make(MapOfColumnsRowsAndColumns.fromMap(map));
+
+    FramedOnHeapAggregatable agger = FramedOnHeapAggregatable.fromRAC(rac);
+
+    final RowsAndColumns results = agger.aggregateAll(
+        frame,
+        new AggregatorFactory[] {
+            new LongSumAggregatorFactory("res", "c2")
+        }
+    );
+
+    new RowsAndColumnsHelper()
+    .expectColumn("c1", c1Vals)
+    .expectColumn("c2", c2Vals)
+    .expectColumn("res", resVals)
+        .allColumnsRegistered()
+        .validate(results);
+  }
```

Comment:
It might be good to also have some tests using the HLL aggregators.  I expect them to be a good thing that exercises the whole "the object gets mutated after `.get()`" thing that I mentioned in a different comment.

Reply:
I've added one - and it just passed because it does a `copy` in the [get()](https://github.com/apache/druid/blob/dff5bcb0a6c693e6fc8131e91c5029dbab132fa1/extensions-core/datasketches/src/main/java/org/apache/druid/query/aggregation/datasketches/hll/HllSketchBuildAggregator.java#L64)